### ISSUE TITLE
Curiosity on the feedforward neural network

### ISSUE THREAD

**Issue by sufengniu** [jwyyy/AdaFNN#1]
Hello @jwyyy thank you for sharing your implementation, which is an interesting paper. Just a few questions regarding to the implementation:
1. I found that your feedforward neural network definition is linear transformation + layer norm with skip connection, activation function with dropout, is there any reasons to do in this way? I mean I often saw people doing skip connection with dense layer, layer norm, etc. but not common to see skip connection on a single layer norm. I just try to understand why this config works, since I did try to remove/modify skip connection, but performance become worse
2. code block 5 for _l1 function, there is a typo to construct torch.ones, should be torch.ones((B, J))
3. I found in the example results, when I add regularization term, the performance dropped heavily. Do you have any insights? I think in the paper results section, it also show that the proposed two regularization terms seems to inhibit the model performance. or if I need to tune them?

**Comment by jwyyy** [comment]
Hi @sufengniu, thank you for your interest in the implementation. Here are my thoughts on your questions.

>     1. I found that your feedforward neural network definition is linear transformation + layer norm with skip connection, activation function with dropout, is there any reasons to do in this way? I mean I often saw people doing skip connection with dense layer, layer norm, etc. but not common to see skip connection on a single layer norm. I just try to understand why this config works, since I did try to remove/modify skip connection, but performance become worse

The feedforward network (FFN) was built specifically for the basis node networks. Since the subsequent network is also a FFN, we keep using this module. The skipping was introduced to help learn basis node networks. Although the overall architecture seems straightforward, it is actually a network-inside-network structure. Making it harder is how these basis nodes get involved in the computational graph through numerical integration. Empirically we found learning this basis networks using dense FFN was not efficient due to two possible reasons: (1) the scale (i.e. the actual magnitude) of the basis function represented by a network can vary a lot, which causes instability; (2) basis node networks need to be of certain size to capture the curvature of target signal functions (in the experiment they are even larger than the subsequent network). For (1), the basis node output is normalized before going to the following layers. For (2), by adding skipping connections, (conceptually) we allow gradient of different levels (internal layers of basis networks) to flow back to early layers. (For example, adding a skipping connection from input layer directly to the output is kinda also aggregating gradients from an identity approximation.) One-layer skipping isn't the only way to do it, and you can do other combinations. 

>     2. code block 5 for _l1 function, there is a typo to construct torch.ones, should be torch.ones((B, J))

Thanks for catching this! Just fixed it. 

>     3. I found in the example results, when I add regularization term, the performance dropped heavily. Do you have any insights? I think in the paper results section, it also show that the proposed two regularization terms seems to inhibit the model performance. or if I need to tune them?

The regularization term means restricting fitted basis nodes. However, restricting the shape of these basis nodes doesn't mean better results. It depends a lot on your data, if the target signal in the data doesn't have these properties, the proposed regularization may not be helpful. In the simulations, two target signal functions do have orthogonal property and have zero-value region. The L1 and L2 penalties are used to help detect these structures / shapes. When using it, it is better to start with more basis nodes and no penalty. If fitted basis networks appear to have similar pattern or zero-value region, then consider reducing the number of basis nodes and also adding small regularization terms.

**Comment by sufengniu** [comment]
make sense, thank you, will close the ticket
